docs/examples/contour_sgld.md

```diff
@@ -0,0 +1,303 @@
+ax.set_xlabel(f"Iterations (x{thinning_factor})")
+ax.set_ylabel("X")
+
+ax.spines["top"].set_visible(False)
+ax.spines["right"].set_visible(False)
+
+
+# Histogram
+ax = plt.subplot(G[0, 2])
+ax.hist(sgld_sample_list, 100)
+ax.set_xlabel("X")
+ax.set_xlim(left=-15, right=35)
+
+ax.set_yticks([])
+ax.spines["top"].set_visible(False)
+ax.spines["right"].set_visible(False)
+ax.spines["left"].set_visible(False)
+
+
+plt.suptitle("Stochastic gradient Langevin dynamics (SGLD)")
+```
+
+```{code-cell} ipython3
+from blackjax.sgmcmc.csgld import CSGLDState
+
+# specify hyperparameters (zeta and sz are the only two hyperparameters to tune)
+zeta = 2
+sz = 10
+temperature = 50
+
+lr = 1e-3
+init_position = 10.0
+
+
+# The following parameters partition the energy space and no tuning is needed.
+num_partitions = 100000
+energy_gap = 0.25
+domain_radius = 50  # restart sampling when the particle explores too deep over the tails and leads to nan.
+
+
+logdensity_fn, grad_fn = gradients.logprob_and_grad_estimator(
+    logprior_fn, loglikelihood_fn, data_size
+)
+csgld = blackjax.csgld(
+    logdensity_fn,
+    grad_fn,
+    zeta=zeta,  # can be specified at each step in lower-level interface
+    temperature=temperature,  # can be specified at each step
+    num_partitions=num_partitions,  # cannot be specified at each step
+    energy_gap=energy_gap,  # cannot be specified at each step
+    min_energy=0,
+)
+
+# 3.1 Simulate via the CSGLD algorithm
+state = csgld.init(init_position)
+
+re_start_counter = 0  # in case the particle goes beyond the domain
+csgld_sample_list, csgld_energy_idx_list = jnp.array([]), jnp.array([])
+
+pb = progress_bar(range(total_iter))
+for iter_ in pb:
+    rng_key, subkey = jax.random.split(rng_key)
+    stepsize_SA = min(1e-2, (iter_ + 100) ** (-0.8)) * sz
+
+    data_batch = jax.random.shuffle(rng_key, X_data)[:batch_size, :]
+    state = jax.jit(csgld.step)(subkey, state, data_batch, lr, stepsize_SA)
+    if jnp.abs(state.position) > domain_radius:
+        re_start_counter += 1
+        state = CSGLDState(
+            jax.random.uniform(rng_key, minval=-domain_radius, maxval=domain_radius),
+            state.energy_pdf,
+            state.energy_idx,
+        )
+
+    if iter_ % thinning_factor == 0:
+        csgld_sample_list = jnp.append(csgld_sample_list, state.position)
+        csgld_energy_idx_list = jnp.append(csgld_energy_idx_list, state.energy_idx)
+        pb.comment = (
+            f"| position {state.position: .2f} | re-restarts {re_start_counter}"
+        )
+```
+
+Contour SGLD takes inspiration from the Wang-Landau algorithm to learn the density of states of the model at each energy level, and uses this information to flatten the target density to be able to explore it more easily.
+
+As a result, the samples returned by contour SGLD are not from the target density directly, and we need to resample them using the density of state as importance weights to get samples from the target distribution.
```

Comment:
Do we expect to do importance sampling like this more than once? Is this worth factoring into a library function?

Reply:
Currently just once. Maybe we can think about this in the future. My current implementation is only acceptable and may not be fast enough, in the future version, we need to include some additional variables showing the important weight for each simulated sample.